components/server/src/analytics-controller.ts

```diff
@@ -0,0 +1,180 @@
+/**
+ * Copyright (c) 2023 Gitpod GmbH. All rights reserved.
+ * Licensed under the GNU Affero General Public License (AGPL).
+ * See License.AGPL.txt in the project root for license information.
+ */
+import { inject, injectable } from "inversify";
+import express from "express";
+import {
+    IAnalyticsWriter,
+    IdentifyMessage,
+    RemoteIdentifyMessage,
+    RemotePageMessage,
+    RemoteTrackMessage,
+} from "@gitpod/gitpod-protocol/lib/analytics";
+import { createCookielessId, maskIp } from "./analytics";
+import { ClientHeaderFields, toClientHeaderFields } from "./express-util";
+import { Config } from "./config";
+import { RateLimited } from "./api/rate-limited";
+import { RateLimitter } from "./rate-limitter";
+import { RateLimiterRes } from "rate-limiter-flexible";
+
+@injectable()
+export class AnalyticsController {
+    constructor(
+        @inject(IAnalyticsWriter) private readonly analytics: IAnalyticsWriter,
+        @inject(Config) private readonly config: Config,
+        @inject(RateLimitter) private readonly rateLimitter: RateLimitter,
+    ) {}
+
+    get router(): express.Router {
+        const router = express.Router();
+        router.post("/trackEvent", async (req, res) => {
+            try {
+                if (await this.rateLimitted(req.user?.id, "trackEvent", res)) {
+                    return;
+                }
+                const clientHeaderFields = toClientHeaderFields(req);
+                const event = req.body as RemoteTrackMessage;
+                this.trackEvent(req.user?.id, event, clientHeaderFields);
+                res.sendStatus(200);
+            } catch (e) {
+                console.error("failed to track event", e);
+                res.sendStatus(500);
+            }
+        });
+        router.post("/trackLocation", async (req, res) => {
+            try {
+                if (await this.rateLimitted(req.user?.id, "trackLocation", res)) {
+                    return;
+                }
+                const clientHeaderFields = toClientHeaderFields(req);
+                const event = req.body as RemotePageMessage;
+                this.trackLocation(req.user?.id, event, clientHeaderFields);
+                res.sendStatus(200);
+            } catch (e) {
+                console.error("failed to track location", e);
+                res.sendStatus(500);
+            }
+        });
+        router.post("/identifyUser", async (req, res) => {
+            try {
+                if (!req.user?.id) {
+                    res.sendStatus(401);
+                    return;
+                }
+                if (await this.rateLimitted(req.user?.id, "identifyUser", res)) {
+                    return;
+                }
+                const clientHeaderFields = toClientHeaderFields(req);
+                const event = req.body as RemoteIdentifyMessage;
```

Comment:
Should we validate body? or ignore them all since we have catched all errors with 500

Reply:
yeah, 500